### PR TITLE
BUILD-57: introduce share projectedresource api

### PR DESCRIPTION
**Pull request by gabemontero**
this is just the API, CRD, and generated client changes for Shares

I'd like to break things up and get this reviewed, agreed upon, merged before moving on to adding the controller and then
address "Access to the custom resource objects can be gated by standard k8s roles and cluster roles." via SARs to see if 
the SA has read access to the Secret or ConfigMap noted in the Share.

/assign @adambkaplan 
@coreydaley FYI

### CODE REVIEW

**Comment by gabemontero** [comment]
Please ignore @adambkaplan @coreydaley 

I made the new api namespace scoped vs. cluster scoped per https://github.com/openshift/builder/tree/master/cmd

I'll update when that is corrected

**Comment by gabemontero** [comment]
OK PR updated @adambkaplan @coreydaley 

The API and client were already clustered.  Now the CRD is clustered instead of namespaces as well.

File: Makefile [hunk 1]
```diff
@@ -8,20 +8,23 @@ REPOSITORY ?= openshift
...
 test: ## Run unit tests. Example: make test
-	go test $(GO_FILES)
+	go test ./cmd/... ./pkg/...
 .PHONY: test
 
 verify: ## Run verifications. Example: make verify
-	go vet $(GO_FILES)
-	gofmt -w cmd/ pkg/
+	go vet ./cmd/... ./pkg/...
+	gofmt -w ./cmd/ ./pkg/
```
Comment (coreydaley) [inline Makefile:30]: I wonder if the "make verify" should also run the "make generate" and verify that the generated files are up to date in the pull request?  I'm not sure if there is a precedent for that in other repos offhand though.

Comment (gabemontero) [reply Makefile:30]: Yeah I saw some stuff in openshift/api that might have been applicable if I had ended up pulling that stuff in:
- https://github.com/openshift/api/blob/master/Makefile#L40
- https://github.com/openshift/api/blob/master/Makefile#L43

The deepcopy verify appears to just pass in an additional env var into the existing deepcopy-gen command.  The CRD one pulls down a dev's personal `yq` release and then seems to only do a couple of very basic checks.  I wasn't as enamored with that one at first blush.

Both are things we could pull in but I ended up taking a "meets min" approach with this PR.

Maybe another opportunity for you and me @coreydaley to divde and conquer after this PR merges, where I move on to adding the controller piece to his work (which should not affect the Makefile), while then you wrap up the cobra piece after this merges and then take a stab at verifies?

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+type ShareList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Share `json:"items"`
+}
+
+// ShareSpec defines the desired state of Share
+type ShareSpec struct {
+	// BackingResource captures the ConfigMap or Secret that is shared.
+	BackingResource `json:"backingResource,omitempty"`
```
Comment (adambkaplan) [inline pkg/api/projectedresource/v1alpha1/share_type.go:32]: This should be required (i.e. no omitempty).

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	// Conditions are the set of k8s Condition instances provided by the associated controller for Shares.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type BackingResource struct {
+	// Kind is a string value representing the REST resource this object represents.
+	// Currently only Secret and ConfigMap are accepted.
+	// +optional
+	Kind string `json:"kind,omitempty"`
```
Comment (adambkaplan) [inline pkg/api/projectedresource/v1alpha1/share_type.go:50]: Required?

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	// APIVersion defines the versioned schema of this representation of an object.
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
```
Comment (adambkaplan) [inline pkg/api/projectedresource/v1alpha1/share_type.go:54]: OK with optional - I believe `""` assumes `core/v1`

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	// Name is the name of the object serving as the backing resource
+	// +optional
+	Name string `json:"name,omitempty"`
```
Comment (adambkaplan) [inline pkg/api/projectedresource/v1alpha1/share_type.go:58]: Required?

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	// Namespace is the namespace of the object serving as the backing resource
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
```
Comment (adambkaplan) [inline pkg/api/projectedresource/v1alpha1/share_type.go:62]: Required?

File: Makefile (hunk 1 above)
Comment (coreydaley) [reply Makefile:30]: Sounds like a plan to me.

File: hack/boilerplate.go.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/*
+Copyright 2020 The OpenShift authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply hack/boilerplate.go.txt:4]: yeah this was generated by the operator sdk while I was still giving that a go

part of next push

File: pkg/api/projectedresource/v1alpha1/share_type.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+// ShareSpec defines the desired state of Share
+type ShareSpec struct {
+	// BackingResource captures the ConfigMap or Secret that is shared.
+	BackingResource `json:"backingResource,omitempty"`
+
+	// Description is a user readable explanation of what the backing resource
+	// provides.
+	Description string `json:"description,omitempty"`
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/api/projectedresource/v1alpha1/share_type.go:36]: part of next push

File: pkg/api/projectedresource/v1alpha1/share_type.go (hunk 2 above)
Comment (gabemontero) [reply pkg/api/projectedresource/v1alpha1/share_type.go:32]: yep

File: pkg/api/projectedresource/v1alpha1/share_type.go (hunk 3 above)
Comment (gabemontero) [reply pkg/api/projectedresource/v1alpha1/share_type.go:50]: yep

File: pkg/api/projectedresource/v1alpha1/share_type.go (hunk 5 above)
Comment (gabemontero) [reply pkg/api/projectedresource/v1alpha1/share_type.go:58]: yep

File: pkg/api/projectedresource/v1alpha1/share_type.go (hunk 6 above)
Comment (gabemontero) [reply pkg/api/projectedresource/v1alpha1/share_type.go:62]: yep

**Comment by gabemontero** [comment]
all your suggestions pushed @adambkaplan 

squashed into https://github.com/openshift/csi-driver-projected-resource/pull/15/commits/44e8a7753500a38a6ba8bbe70259a863d3cb516e

**Comment by coreydaley** [comment]
@adambkaplan bump

**Pull request merged**